pkg/graveler/graveler.go

```diff
@@ -2039,6 +2055,61 @@ func (g *Graveler) deleteUnsafe(ctx context.Context, repository *RepositoryRecor
 	return nil
 }
 
+// deleteNoTombstoneUnsafe will try to remove the key from KV without adding a tombstone if possible, under the following conditions:
+// 1. Key is not in committed data
+// 2. Key does not exist in any sealed tokens
+// In any other case, a tombstone will be created. This flow ignores the delete sensor and compaction mechanism.
+func (g *Graveler) deleteNoTombstoneUnsafe(ctx context.Context, repository *RepositoryRecord, key Key, branchRecord BranchRecord) error {
+	// check key in committed - do we need tombstone?
+	var metaRangeID MetaRangeID
+	if branchRecord.CompactedBaseMetaRangeID != "" {
+		metaRangeID = branchRecord.CompactedBaseMetaRangeID
+	} else {
+		commit, err := g.RefManager.GetCommit(ctx, repository, branchRecord.CommitID)
+		if err != nil {
+			return err
+		}
+		metaRangeID = commit.MetaRangeID
+	}
+
+	_, err := g.CommittedManager.Get(ctx, repository.StorageID, repository.StorageNamespace, metaRangeID, key)
+	if err == nil {
+		// found in committed, set tombstone
+		return g.deleteAndNotify(ctx, repository.RepositoryID, branchRecord, key, false)
+	}
+	if !errors.Is(err, ErrNotFound) {
+		// unknown error
+		return fmt.Errorf("reading from committed: %w", err)
+	}
+
+	// key is not committed - check sealed tokens
+	// find first key in sealed tokens, if found based on value
+	if branchRecord.StagingToken == "" {
+		return fmt.Errorf("missing staging token: %w", ErrNotFound)
+	}
+	for _, st := range branchRecord.SealedTokens {
+		value, err := g.StagingManager.Get(ctx, st, key)
+		// iterate to next sealed token if key not found
+		if errors.Is(err, ErrNotFound) {
+			continue
+		}
+		// unknown error
+		if err != nil {
+			return err
+		}
+		// Found in sealed
+		if value == nil {
+			// nothing to do, latest value is already a tombstone
+			return nil
+		}
```

Comment:
Just to check, our sealed tokens are sorted, so if it was last deleted, you will first see the tombstone, right?